lxd/project/permissions.go

```diff
@@ -1465,22 +1465,71 @@ var aggregateLimitConfigValuePrinters = map[string]func(int64) string{
 	},
 }
 
-// FilterUsedBy filters a UsedBy list based on project access.
+// FilterUsedBy filters a UsedBy list based on the entities that the requestor is able to view.
 func FilterUsedBy(authorizer auth.Authorizer, r *http.Request, entries []string) []string {
-	// Filter the entries.
-	usedBy := []string{}
+	// Get a map of URLs by entity type. If there are multiple entries of a particular entity type we can reduce the
+	// number of calls to the authorizer.
+	urlsByEntityType := make(map[entity.Type][]*api.URL)
 	for _, entry := range entries {
 		u, err := url.Parse(entry)
 		if err != nil {
 			logger.Warn("Failed to parse project used-by entity URL", logger.Ctx{"url": entry, "error": err})
+			continue
+		}
+
+		entityType, projectName, location, pathArguments, err := entity.ParseURL(*u)
+		if err != nil {
+			logger.Warn("Failed to parse project used-by entity URL", logger.Ctx{"url": entry, "error": err})
+			continue
 		}
 
-		err = authorizer.CheckPermission(r.Context(), r, &api.URL{URL: *u}, auth.EntitlementCanView)
+		entityURL, err := entityType.URL(projectName, location, pathArguments...)
 		if err != nil {
+			logger.Warn("Failed to create canonical entity URL for project used-by filtering", logger.Ctx{"url": entry, "error": err})
+			continue
+		}
+
+		urlsByEntityType[entityType] = append(urlsByEntityType[entityType], entityURL)
+	}
+
+	// Filter the entries.
+	usedBy := make([]string, 0, len(entries))
+
+	// Used-by lists do not include the project query parameter if it is the default project.
+	appendUsedBy := func(u *api.URL) {
+		if u.Query().Get("project") == api.ProjectDefaultName {
+			q := u.Query()
+			q.Del("project")
+			u.RawQuery = q.Encode()
+		}
+
+		usedBy = append(usedBy, u.String())
+	}
+
+	for entityType, urls := range urlsByEntityType {
+		// If only one entry of this type, check directly.
+		if len(urls) == 1 {
+			err := authorizer.CheckPermission(r.Context(), r, urls[0], auth.EntitlementCanView)
```

Comment:
WRT to

> The given UsedBy URLs may not contain the "project" query parameter.
> Passing them directly to the authorizer is problematic because the OpenFGA datastore depends on this.

This seems like a future bug waiting to happen elsewhere as this is making it every caller's job to rewrite every URL to ensure it always has a project param where appropriate, rather than the authorizer handling it centrally.

Is there a reason you're doing it this way, and can we move the logic into the authorizer?

I wont block the PR on this - but think we should add to the refactoring tasks to make this less brittle going forward.

Reply:
If the authorizer handled it internally then we wouldnt also need to strip the project again later either.

Reply:
Yeah that makes sense. I'll create an issue.